wikipedia/track.py

```diff
@@ -0,0 +1,108 @@
+
+        queries = [query for query, _ in queries_with_probabilities]
+        probabilities = [float(probability) for _, probability in queries_with_probabilities]
+        random.seed(random_seed)
+        for query in random.choices(queries, weights=probabilities, k=k):
+            # remove special chars from the query + lowercase
+            yield QUERY_CLEAN_REXEXP.sub(" ", query).lower()
+
+
+class SearchApplicationParams:
+    def __init__(self, track, params):
+        self.indices = params.get("indices", track.index_names())
+        self.name = params.get("search-application", f"{self.indices[0]}-search-application")
+
+
+class CreateSearchApplicationParamSource(ParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self.search_application_params = SearchApplicationParams(track, params)
+
+    def partition(self, partition_index, total_partitions):
+        return self
+
+    def params(self):
+        return {
+            "method": "PUT",
+            "path": f"{SEARCH_APPLICATION_ROOT_ENDPOINT}/{self.search_application_params.name}",
+            "body": {"indices": self.search_application_params.indices},
+        }
+
+
+class QueryIteratorParamSource(ParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self._queries_iterator = None
+
+    def size(self):
+        return self._params.get("iterations", 10000)
+
+    def partition(self, partition_index, total_partitions):
+        if self._queries_iterator is None:
+            partition_size = math.ceil(self.size() / total_partitions)
+            self._queries_iterator = query_iterator(partition_size, random_seed=self._params.get("seed", None))
+        return self
+
+
+class SearchApplicationSearchParamSource(QueryIteratorParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self.search_application_params = SearchApplicationParams(track, params)
+
+    def params(self):
+        query = next(self._queries_iterator)
+        return {
+            "method": "POST",
+            "path": f"{SEARCH_APPLICATION_ROOT_ENDPOINT}/{self.search_application_params.name}/_search",
+            "body": {
+                "params": {
+                    "query_string": query,
+                },
+            },
+        }
+
+
+class QueryParamSource(QueryIteratorParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self._index_name = params.get("index", track.indices[0].name if len(track.indices) == 1 else "_all")
+        self._cache = params.get("cache", True)
+
+    def params(self):
+        result = {
+            "body": {"query": {"query_string": {"query": next(self._queries_iterator), "default_field": self._params["search-fields"]}}},
```

Comment:
Let's use a multi-match query to avoid the reserved keywords (and, or, ...)?

Reply:
I will do this in a follow-up. We still have few run to do and I would prefer to keep the same query for now to get comparable results.